File: bench/bytebeat/bytebeat.go
package main

import (
	"fmt"
	"time"
)

const (
	n       = 1 << 21
	nRuns   = 21
	nWarmup = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func checksumU8(out []uint8) uint32 {
	h := uint32(0x811c9dc5)
	for _, b := range out {
		h = mix(h, uint32(b))
	}
	return h
}

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

// Go binds *, <<, >>, & at the same (highest) precedence, unlike C/JS — so the
// voice groupings are parenthesized explicitly to match the JS source's meaning.
func sample(t uint32) uint32 {
	v1 := ((t * 5) & (t >> 7)) | ((t * 3) & (t >> 10))
	v2 := t * (((t >> 12) | (t >> 8)) & (63 & (t >> 4)))
	return (v1 + v2) & 255
}

func render(buf []uint8) {
	for t := range buf {
		buf[t] = uint8(sample(uint32(t)))
	}
}

func main() {
	buf := make([]uint8, n)
	for i := 0; i < nWarmup; i++ {
		render(buf)
	}
	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		render(buf)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), checksumU8(buf), n, 1, nRuns)
}

File: bench/bytebeat/bytebeat.rs
use std::time::Instant;

const N: usize = 1 << 21;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn checksum_u8(out: &[u8]) -> u32 {
    let mut h = 0x811c_9dc5u32;
    for &b in out {
        h = mix(h, b as u32);
    }
    h
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

#[inline]
fn sample(t: u32) -> u32 {
    let v1 = (t.wrapping_mul(5) & (t >> 7)) | (t.wrapping_mul(3) & (t >> 10));
    let v2 = t.wrapping_mul(((t >> 12) | (t >> 8)) & (63 & (t >> 4)));
    v1.wrapping_add(v2) & 255
}

fn render(buf: &mut [u8]) {
    for (t, slot) in buf.iter_mut().enumerate() {
        *slot = sample(t as u32) as u8;
    }
}

fn main() {
    let mut buf = vec![0u8; N];
    for _ in 0..N_WARMUP {
        render(&mut buf);
    }

    let mut samples = [0.0; N_RUNS];
    for sample in &mut samples {
        let t0 = Instant::now();
        render(&mut buf);
        *sample = t0.elapsed().as_secs_f64() * 1000.0;
    }

    let us = median_us(&mut samples);
    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        us,
        checksum_u8(&buf),
        N,
        1,
        N_RUNS
    );
}
